add module that implements a bias operation

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64296

This module is going to be a game changer. Let's try to get it in
before 1.10.

Differential Revision: [D30700670](https://our.internmc.facebook.com/intern/diff/D30700670)

Commit:
add module that implements a bias operation

This module is going to be a game changer. Let's try to get it in
before 1.10.

[ghstack-poisoned]

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

[ghstack-poisoned]

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

[ghstack-poisoned]

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

[ghstack-poisoned]

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30700670).

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

Differential Revision: [D30700670](https://our.internmc.facebook.com/intern/diff/D30700670)

[ghstack-poisoned]

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

Differential Revision: [D30700670](https://our.internmc.facebook.com/intern/diff/D30700670)

[ghstack-poisoned]

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -1829,6 +1829,24 @@ def hardsigmoid(input: Tensor, inplace: bool = False) -> Tensor:
+
+    Shape:
+
+        - Input: :math:`(N, *, num\_features)` N is the batch size, `*` means any number of
+          additional dimensions
+        - Bias: :math:`(num\_features)`
+        - Output: :math:`(N, *, num\_features)`
```
nit: shapes here should match documented shapes for the module form.

I personally like `(*, num_features), where * is any number of dimensions, including none` better because I think it's more accurate, as e.g. `torch.randn(num_features)` should be a valid input. When the `N` is present, it makes it seem required

dagitses replied:
looks good to me!

jbschlosser commented on test/cpp/api/functional.cpp:
```diff
@@ -1248,6 +1248,24 @@ TEST_F(FunctionalTest, LocalResponseNorm) {
   ASSERT_TRUE(torch::allclose(y, y_exp, 1e-4, 1e-7));
 }
 
+TEST_F(FunctionalTest, Bias) {
+  const auto input = torch::tensor({2.0, 3.0});
+  const auto bias = torch::tensor({5.0, 7.0});
+  const auto result = F::bias(input, bias);
+  const auto expected = torch::tensor({7.0, 10.0});
+  torch::test::assert_tensor_equal(result, expected);
+}
+
+TEST_F(FunctionalTest, BiasBroadcast) {
+  const auto input = torch::tensor({{2.0, 3.0},
+                                   {5.0, 7.0}});
+  const auto bias = torch::tensor({11.0, 13.0});
+  const auto result = F::bias(input, bias);
+  const auto expected = torch::tensor({{13.0, 16.0},
+                                      {16.0, 20.0}});
+  torch::test::assert_tensor_equal(result, expected);
+}
```
flawless!

dagitses replied:
😉

jbschlosser commented on test/cpp/api/modules.cpp:
```diff
@@ -1060,6 +1061,22 @@ TEST_F(ModulesTest, MaxPool3d_MaxUnpool3d) {
+  torch::Tensor s = output.sum();
+
+  s.backward();
+  ASSERT_THAT(output.sizes(), testing::ElementsAre(2));
```
Alternatively to using gmock for this assert, you could do what the previous test does:
```
ASSERT_EQ(output.sizes(), std::vector<int64_t>({2}));
```

dagitses replied:
Oh nice. My memory told me that this used to give a poor output, now it's much better.

```
  auto sizes = std::vector<int64_t>{3, 5, 7, 11, 13};
  EXPECT_EQ(sizes, (std::vector<int64_t>{5, 7}));
  EXPECT_THAT(sizes, testing::ElementsAre(5, 7));
```

`EXPECT_EQ`:
```
Expected equality of these values:
  sizes
    Which is: { 3, 5, 7, 11, 13 }
  (std::vector<int64_t>{5, 7})
    Which is: { 5, 7 }
```

`EXPECT_THAT`
```
Value of: sizes
Expected: has 2 elements where
element #0 is equal to 5,
element #1 is equal to 7
  Actual: { 3, 5, 7, 11, 13 }, which has 5 elements
```

Very comparable! Advantages to `ElementsAre`: you get counts and you have an explicit distinction between expected and actual. But that's not that compelling. I'll change for this PR to be consistent.

jbschlosser commented on test/cpp/api/CMakeLists.txt:
```diff
@@ -51,6 +51,7 @@ endif()
 
 add_executable(test_api ${TORCH_API_TEST_SOURCES})
 target_include_directories(test_api PRIVATE ${ATen_CPU_INCLUDE})
+target_link_libraries(test_api PRIVATE torch gmock)
```
Interesting that there's no usage of gmock already in the codebase!

dagitses replied:
There's a few, just not in the api tests here.

```
$ git grep gmock | grep --invert-match ^third_party/
aten/src/ATen/core/boxing/impl/test_helpers.h:#include <gmock/gmock.h>
aten/src/ATen/test/Dict_test.cpp:#include <gmock/gmock.h>
aten/src/ATen/test/ivalue_test.cpp:#include <gmock/gmock.h>
aten/src/ATen/test/vitals.cpp:#include <gmock/gmock.h>
c10/test/CMakeLists.txt:    target_link_libraries(${test_name} c10 gmock gtest gtest_main)
c10/test/util/either_test.cpp:#include <gmock/gmock.h>
c10/test/util/optional_test.cpp:#include <gmock/gmock.h>
c10/test/util/string_view_test.cpp:#include <gmock/gmock.h>
cmake/Dependencies.cmake:  set(BUILD_GMOCK ON CACHE BOOL "Build gmock." FORCE)
test/cpp/api/CMakeLists.txt:target_link_libraries(test_api PRIVATE torch gmock)
test/cpp/api/modules.cpp:#include <gmock/gmock.h>
test/cpp/jit/test_interpreter.cpp:#include <gmock/gmock.h>
```

jbschlosser commented on torch/csrc/api/include/torch/nn/functional/linear.h:
```diff
@@ -12,6 +12,12 @@ inline Tensor bilinear(const Tensor& input1, const Tensor& input2, const Tensor&
 // ============================================================================
 
+inline Tensor bias(const Tensor& input, const Tensor& bias) {
+  return torch::bias(input, bias);
```
Very nice, you resisted the temptation to just call add here :)

dagitses replied:
it's not too late. i could even assert that it has the same behavior as calling the bias operator.

jbschlosser commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -25,6 +25,31 @@ Tensor IdentityImpl::forward(const Tensor& input) {
+}
+
+void BiasImpl::reset_parameters() {
+  const auto bound = 1 / std::sqrt(options.num_features());
+  torch::nn::init::uniform_(bias, -bound, bound);
```
nit: should initialize from standard normal distribution (`torch::nn::init::normal_`)

dagitses replied:
done. what's the rationale? note that linear initializes its bias from uniform distribution.

what is a sensible stddev for this?

jbschlosser replied:
Rationale is just that the lab claims it should arbitrarily :p No special reason beyond that really. "The module should accept a constructor argument named num_features that defines the size of its vector parameter, and the vector should be randomly initialized from a standard normal distribution."

> what is a sensible stddev for this?

Standard normal implies 0 mean and 1 stddev

dagitses replied:
oops. my bad!

jbschlosser commented on torch/nn/modules/linear.py:
```diff
@@ -108,6 +108,48 @@ def extra_repr(self) -> str:
+        self.num_features = num_features
+        self.bias = Parameter(torch.empty(num_features, device=device,
+                                          dtype=dtype))
+        init.uniform_(self.bias)
```
nit: should be `init.normal_`

dagitses replied:
done!

jbschlosser reviewed (commented):
Great work! Left a few minor comments below

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

Differential Revision: [D30700670](https://our.internmc.facebook.com/intern/diff/D30700670)

[ghstack-poisoned]

dagitses reviewed (commented):
thanks!

Commit:
Update on "add module that implements a bias operation"


This module is going to be a game changer. Let's try to get it in
before 1.10.

Differential Revision: [D30700670](https://our.internmc.facebook.com/intern/diff/D30700670)

[ghstack-poisoned]

jbschlosser reviewed (approved):
LGTM!

jbschlosser commented:
Closing as complete!
